### PR TITLE
Multiconfig

### PR DESCRIPTION
**Pull request by dmtrio**

### CODE REVIEW

File: appAuthHelperFetchTokens.js [hunk 1]
```diff
@@ -5,41 +5,67 @@
 
     // appAuth expects the details to be provided via hash, so copy them there
     window.location.hash = window.location.search.substring(1); // removes the '?'
+    var appAuthConfig;
 
-    var appAuthConfig = JSON.parse(localStorage.getItem("appAuthConfig")),
-        TokenManager = require("./TokenManager"),
-        tokenManager;
+    if (window.location.search) {
+        const params = new URLSearchParams(window.location.search);
```
Comment (jakefeasel) [inline appAuthHelperFetchTokens.js:11]: Unfortunately this function isn't available in Internet Explorer :(

File: TokenManager.js [hunk 2]
```diff
@@ -117,6 +121,7 @@
          * Helper function that reduces the amount of duplicated code, as there are several different
          * places in the code that require initiating an authorization request.
          */
+        //NOTE: Local to TOKEN MANAGER
```
Comment (jakefeasel) [inline TokenManager.js:124]: Combine with above comment block

File: TokenManager.js [hunk 3]
```diff
@@ -312,15 +316,17 @@
                         return Promise.reject("invalid_token");
                     }
                 })
-                .then((request) => fetch(request.url, request.options))
+                .then((request) => {
+                    return fetch(request.url, request.options);})
```
Comment (jakefeasel) [inline TokenManager.js:320]: Why make this change?

File: TokenManager.js [hunk 4]
```diff
@@ -312,15 +316,17 @@
...
                 .then((response) => {
                     if (this.getAuthHeaderDetails(response)["error"] === "invalid_token") {
                         return Promise.reject("invalid_token");
                     } else {
                         return response;
                     }
                 })
-                .then((response) => this.serializeResponse(response));
+                .then((response) => {
+                    return this.serializeResponse(response);});
```
Comment (jakefeasel) [inline TokenManager.js:329]: Also - why make this one?

File: TokenManager.js [hunk 5]
```diff
@@ -338,14 +344,16 @@
...
+            return response.text().then((bodyText) => {
+                return {
+                    body: bodyText,
+                    init: {
+                        status: response.status,
+                        statusText: response.statusText,
+                        headers: this.serializeHeaders(response.headers)
+                    }
+                };
+            });
```
Comment (jakefeasel) [inline TokenManager.js:356]: I'm detecting a trend - you don't like implicit returns from a fat arrow function? :)

File: appAuthHelper.js [hunk 6]
```diff
@@ -144,41 +103,107 @@
...
-            document.getElementsByTagName("body")[0].appendChild(this.rsIframe);
+            document.getElementsByTagName("body")[0].appendChild(appConfigs[authId].appAuthIframe);
+            document.getElementsByTagName("body")[0].appendChild(appConfigs[authId].rsIframe);
+
+            // must be this because only one proxy exists
+            appConfigs[authId].registerIdentityProxy = this.registerIdentityProxy;
+            appConfigs[authId].registerXHRProxy = this.registerXHRProxy.bind(this);
+            appConfigs[authId].renewTokens = this.renewTokens.bind(appConfigs[authId]);
+            appConfigs[authId].whenRenewTokenFrameAvailable = this.whenRenewTokenFrameAvailable.bind(appConfigs[authId]);
+            appConfigs[authId].logout = this.logout;
```
Comment (jakefeasel) [inline appAuthHelper.js:145]: Could you explain why you need these above lines?

File: IdentityProxyCore.js [hunk 7]
```diff
@@ -1,15 +1,41 @@
...
+                resourceServers: resourceServers,
+                transmissionPort: transmissionPort,
+                failedRequestQueue: this.failedRequestQueue || {},
+            };
+        },
+        removeProxyCore: function (hostname) {
+            delete proxyConfigs[hostname];
+        },
         renewTokens: function (resourceServer) {
+            this.setProxyCoreByUrl(resourceServer);
```
Comment (jakefeasel) [inline IdentityProxyCore.js:38]: Could you explain why this is needed at this point in the code? It seems strange to me...

File: IdentityProxyXHR.js [hunk 8]
```diff
@@ -18,6 +18,7 @@
                 if (typeof isAsync !== "undefined" && !isAsync) {
                     throw "Synchronous XHR requests not supported";
                 }
+                _this.setProxyCoreByUrl(url);
```
Comment (jakefeasel) [inline IdentityProxyXHR.js:21]: So the theory here is that the specific set of core details used by the proxy is swapped around as each request is processed? 

BTW it seems that we are missing equivalent changes to the IdentityProxyServiceWorker implementation

File: appAuthHelper.js [hunk 9]
```diff
@@ -279,7 +333,7 @@
                     }).bind(this);
 
                     registerServiceWorker().then((function () {
-                        this.identityProxy = {
+                        identityProxy = {
                             tokensRenewed: function (currentResourceServer) {
```
Comment (jakefeasel) [inline appAuthHelper.js:337]: It appears that we need to add all of the equivalent support on the service worker side of things

**Review by jakefeasel — COMMENTED** [review]
A few things to deal with - service worker support and some questions about the implementation. Could you please also update the README with some details about how this might be used to support multiple IDPs? Thanks!

ps Sorry for the delay in getting to this review

File: TokenManager.js (hunk 3 above)
Comment (dmtrio) [reply TokenManager.js:320]: I'll switch back to implicit returns - it was easier to read with the and know if the return was needed with returns on a different line

**Comment by jakefeasel** [comment]
Pushed to master separately from this PR

**Pull request closed without merging**